### PR TITLE
fix: Using Neo4J transactions

### LINKED ISSUE #78: Link sessions to requests and handle transactions
Right now the initialization of the neo4j session is very rudimentary in the backend.

It is shared among requests and we don't handle transactions.

Having a concurrent application we really need to handle both of that, or this may lead to very inconsistent database states.

The simplest strategy is to have one session and transaction per request. The transaction should be rolled back if an exception pops up, and be committed otherwise (this is the default behavior of the transaction context manager provided by neo4j)

We can do that through a [middleware](https://fastapi.tiangolo.com/tutorial/middleware/) and open [session and transaction](https://neo4j.com/docs/api/python-driver/current/api.html#sessions-transactions) there.

The middleware will use the transaction context manager and call next_call inside it.
It has to store the returned transaction in a [contextvar](https://docs.python.org/3/library/contextvars.html) which is a global variable in the entries.py module.

Instead of using the session in entries.py (and everywhere else), we should instead get the transaction through the contextvar .

### PR DESCRIPTION
**Pull request by aadarsh-ram**
### What
- Using transactions for requests in order to handle them concurrently.

### Related bug(s)
- Fixes #78

### CODE REVIEW

File: backend/editor/api.py
Comment (alexgarel) [inline backend/editor/api.py:66]: Put it in it's own file (with initialize_db and shutdown_db. So that you don't have cross import problems.

Maybe graph_db.py

File: backend/editor/api.py
Comment (alexgarel) [inline backend/editor/api.py:63]: give at better name :-) (I know you are still experimenting) like current_transaction

File: backend/editor/api.py
Comment (alexgarel) [inline backend/editor/api.py:98]: Instead of passing txn around, if you put it in a graph_db.py file you can import it in entries.py and use it there. But beware as it is a global, you should not use:
```python
from .graph_db import current_transaction
...
current_transaction.do_something
```
Because you would miss updates.

You have to instead keep the namespace:

```python
from . import graph_db
...
graph_db.current_transaction.do_something
```

File: backend/editor/entries.py
Comment (alexgarel) [inline backend/editor/entries.py:22]: As said above don't put it neither in entries nor in api, put it in it's own module.

File: backend/editor/api.py
Comment (alexgarel) [inline backend/editor/api.py:66]: You really have to use the context manager ! because here, you never end the transaction !
See https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Transaction

```suggestion
        with session.begin_transaction() as _txn:
            # make it current transaction
            txn = _txn
            response = await call_next(request)
        txn = None
        return response
```

File: backend/editor/api.py
Comment (alexgarel) [inline backend/editor/api.py:62]: If you're interested we can make this part a context manager in graph_db (so that all code managing db is in there):

```python
@contextlib.contextmanager
def TransactionCtx():
    """Transaction context will set global transaction txn for the code in context

    transaction is automatically rollback if an exception occurs within the context
    """
    global txn
    with driver.session() as session
        with session.begin_transaction() as _txn:
            txn = _txn
            yield txn
    txn = None
```

and here:

```python
@app.middleware("http")
async def initialize_neo4j_transactions(request: Request, call_next):
    with graph_db.TransactionCtx():
        response = await call_next(request)
```

File: backend/editor/graph_db.py
Comment (alexgarel) [inline backend/editor/graph_db.py:4]: Ok, now the next step (because we want to support concurrency) is to make it compatible with concurrency (in an async context).

For this, you will need a [context variable](https://docs.python.org/3/library/contextvars.html)

It's better to encapsulate "getting the transaction" in a function. So that in entries.py you have:
`transaction().run(....)`
(note that since you will use a function, you can now import the function directly: `from graph_bd import transaction()`)
This function should return current value of global context variable, or raise if it's not set (or set to None)

**Review by alexgarel — CHANGES_REQUESTED** [review]
Cool, now let's introduce context variables and it will be ready.

**Review by alexgarel — APPROVED** [review]
Great !
Almost there, a small change required. However I approve so that you can merge right after that.

**Comment by aadarsh-ram** [comment]
@alexgarel I think we can merge this after the search functionality PR is merged.

**Comment by aadarsh-ram** [comment]
Thanks for your help @alexgarel! 👍

**Pull request merged**
